scalyr_agent/date_parsing_utils.py

```diff
@@ -0,0 +1,465 @@
+        to_nanos = 9 - len(fractions)
+        nanos = int(int(fractions) * 10 ** to_nanos)
+
+    return nano_seconds + nanos
+
+
+def _rfc3339_to_nanoseconds_since_epoch_regex(string):
+    """
+    rfc3339_to_nanoseconds_since_epoch variation which utilizes regex approach
+    """
+    # split the string in to main time and fractional component
+    parts = string.split(".")
+
+    # it's possible that the time does not have a fractional component
+    # e.g 2015-08-03T09:12:43Z, in this case 'parts' will only have a
+    # single element that should end in Z.  Strip the Z if it exists
+    # so we can use the same format string for processing the main
+    # date+time regardless of whether the time has a fractional component.
+    if parts[0].endswith("Z"):
+        parts[0] = parts[0][:-1]
+
+    # NOTE: We benchmarked multiple versions, including regex one and using string.split
+    # appears to be a bit faster than regex
+    try:
+        dt = datetime.datetime(
+            *list(map(int, RFC3339_STR_REGEX.match(string).groups()))
+        )
+    except Exception:
+        return None
+
+    nano_seconds = (
+        calendar.timegm((dt.year, dt.month, dt.day, dt.hour, dt.minute, dt.second))
+        * 1000000000
+    )
+
+    nanos = 0
+
+    # now add the fractional part
+    if len(parts) > 1:
+        fractions = parts[1]
+        # if the fractional part doesn't end in Z we likely have a
+        # malformed time, so just return the current value
+        if not fractions.endswith("Z"):
+            # we don't handle non UTC timezones yet
+            if any(c in fractions for c in "+-"):
+                return None
+
+            return nano_seconds
+
+        # strip the final 'Z' and use the final number for processing
+        fractions = fractions[:-1]
+        to_nanos = 9 - len(fractions)
+        nanos = int(int(fractions) * 10 ** to_nanos)
+
+    return nano_seconds + nanos
+
+
+def _rfc3339_to_nanoseconds_since_epoch_string_split(string):
+    """
+    rfc3339_to_nanoseconds_since_epoch variation which utilizes string.split approach
+    """
+    # split the string in to main time and fractional component
+    parts = string.split(".")
+
+    # it's possible that the time does not have a fractional component
+    # e.g 2015-08-03T09:12:43Z, in this case 'parts' will only have a
+    # single element that should end in Z.  Strip the Z if it exists
+    # so we can use the same format string for processing the main
+    # date+time regardless of whether the time has a fractional component.
+    if parts[0].endswith("Z"):
+        parts[0] = parts[0][:-1]
+
+    # NOTE: We benchmarked multiple versions, including regex one and using string.split
+    # appears to be a bit faster than regex
+    try:
+        # NOTE: I intentionally access values directly in the list and don't assign them
+        # to intermediate variables since it's faster
+        result = parts[0].split("T")
+        result1 = result[0].split("-")
```

Comment:
For clarity, could we rename `result1` to something like `date_parts` and `result2` to something like `time_parts`?  This will make it easier to read this in 6 months time.

Reply:
That's what I originally did (unpacked it in a tuple and used more self explanatory names), but then I did some testing and noticed it had small overhead and results in tons of short-lived variables so I decided to remove it (see the code comment above).

Reply:
d2cfd72.

Reply:
Sorry, I should have been more clear, I don't mean to unpack them, I just mean rename them, something like:

```
...
date_and_time = parts[0].split("T")
date_parts = date_and_time[0].split("-")
time_parts = date_and_time[1].split(":")
...
```

That way it's clear which variable holds which part of the timestamp.  Not to mention, `result1` and `result[1]` look very similar and are easy to confuse.